sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -599,7 +599,7 @@ pi_result _pi_context::finalize() {
   // For example, zeEventPool could be still alive.
   std::lock_guard<std::mutex> NumEventsLiveInEventPoolGuard(
       NumEventsLiveInEventPoolMutex);
-  if (ZeEventPool && NumEventsLiveInEventPool[ZeEventPool])
+  if (ZeEventPool)
```

Comment:
How is it possible that there are still live events in a pool? Shouldn't they all have be released already? Is this some leak of events indeed?

Reply:
Actually, the problem is that there is an EventPool that is still allocated even if there is NO live events for it. This caused memory leaks.
Previously, we only destroyed `EventPool` only if there are any live events for it, and did not destroy it if no events are live.

The change in this PR destroys any allocated EventPool regardless of live events because it is the clean up process.
It fixed many reported memory leaks in EventPool reported by Valgrind.

Reply:
So, this is a wrong name `NumEventsLiveInEventPool` since it is not counting live events. It counts the number of not released entries in the pool (starts with all entries not released). Please rename it. So the expectation is still that here all events are already released but not all pool entries are released (those that were not used are not released for sure). 

What if the `NumEventsLiveInEventPool` did get to 0, and the L0 pool was already released in `_pi_context::decrementAliveEventsInPool`? Won't it attempt to delete it again, and fail?

Reply:
@jandres742 , 
If we call ZeEventPoolDestroy() for the already destroyed event pool, does it cause a problem or is it ignored silently?

Reply:
> If we call ZeEventPoolDestroy() for the already destroyed event pool, does it cause a problem 

Yes, since when internally we try to dereference the pool's pointer, we will fail with a segfault. If you are using L0's  validation layer, then it will catch the null pointer after destroy and you will get just a ZE_RESULT_ERROR_INVALID_NULL_HANDLE error

Reply:
@jandres742, Thanks for confirmation.
@smaslov-intel, 
I nullified the global pointer ZeEventPool whenever the EnentPool is destroyed.
So, we won't call ZeEventPoolDestroy() for the already destroyed pool again.

Reply:
You are not nullifying the right `ZeEventPool`. There are `ZeEventPool` in both pi_event and pi_context.
Also, please address this comment:

> this is a wrong name NumEventsLiveInEventPool since it is not counting live events. It counts the number of not released entries in the pool (starts with all entries not released). Please rename it.

Reply:
Actually, you are nullifying and checking the same one, in `pi_context`.
However, it is not correct to nullify the one in `pi_context` upon decrement of alive slots in some `p_event`'s pool.

Reply:
If I understand correctly, `pi_context` and `pi_event` are sharing the same pointer to the EventPool.
We need to nullify the pi_context::ZeEventPool because that is the one we use in pi_context::finalize().